drivers/wifi/infineon/airoc_wifi.c

```diff
@@ -0,0 +1,652 @@
+	whd_security_t security;
+	whd_ssid_t ssid = {0};
+	uint8_t channel = 0;
+
+	if (k_sem_take(&data->sema_common, K_MSEC(AIROC_WIFI_WAIT_SEMA_MS)) != 0) {
+		return -EAGAIN;
+	}
+
+	if (data->second_interface_init == 0) {
+		if (whd_add_secondary_interface(data->whd_drv, NULL, &airoc_ap_if) !=
+		    CY_RSLT_SUCCESS) {
+			LOG_ERR("Error Unable to bring up the whd secondary interface");
+			k_sem_give(&data->sema_common);
+			return -EAGAIN;
+		}
+		data->second_interface_init = 1;
+	}
+
+	if (data->is_ap_up) {
+		LOG_ERR("Already AP is on - first disconnect");
+		k_sem_give(&data->sema_common);
+		return -EAGAIN;
+	}
+
+	ssid.length = params->ssid_length;
+	memcpy((void *)ssid.value, (void *)params->ssid, ssid.length);
+
+	if (((params->channel > 0) && (params->channel < 12)) ||
+	    ((params->channel > 35) && (params->channel < 166))) {
+		channel = params->channel;
+	} else {
+		channel = 1;
+	}
+
+	if (params->psk_length == 0) {
+		security = WHD_SECURITY_OPEN;
+	} else {
+		security = WHD_SECURITY_WPA2_AES_PSK;
+	}
+
+	if (whd_wifi_init_ap(airoc_ap_if, &ssid, security, (const uint8_t *)params->psk,
+			     params->psk_length, channel) != 0) {
+		LOG_ERR("Failed to init whd ap interface");
+		k_sem_give(&data->sema_common);
+		return -EAGAIN;
+	}
+
+	if (whd_wifi_start_ap(airoc_ap_if) != 0) {
+		LOG_ERR("Failed to start whd ap interface");
+		k_sem_give(&data->sema_common);
+		return -EAGAIN;
+	}
+
+	/* set event handler */
+	if (whd_management_set_event_handler(airoc_ap_if, ap_link_events, ap_link_events_handler,
+					     NULL, &ap_event_handler_index) != 0) {
+		whd_wifi_stop_ap(airoc_ap_if);
+		k_sem_give(&data->sema_common);
+		return -EAGAIN;
+	}
+
+	data->is_ap_up = 1;
+
+	k_sem_give(&data->sema_common);
+	return 0;
+}
+
+#if defined(CONFIG_NET_STATISTICS_WIFI)
+static int airoc_mgmt_wifi_stats(const struct device *dev, struct net_stats_wifi *stats)
+{
+	struct airoc_wifi_data *data = dev->data;
+
+	stats->bytes.received = data->stats.bytes.received;
+	stats->bytes.sent = data->stats.bytes.sent;
+	stats->pkts.rx = data->stats.pkts.rx;
+	stats->pkts.tx = data->stats.pkts.tx;
+	stats->errors.rx = data->stats.errors.rx;
+	stats->errors.tx = data->stats.errors.tx;
+	stats->broadcast.rx = data->stats.broadcast.rx;
+	stats->broadcast.tx = data->stats.broadcast.tx;
+	stats->multicast.rx = data->stats.multicast.rx;
+	stats->multicast.tx = data->stats.multicast.tx;
+	stats->sta_mgmt.beacons_rx = data->stats.sta_mgmt.beacons_rx;
+	stats->sta_mgmt.beacons_miss = data->stats.sta_mgmt.beacons_miss;
+
+	return 0;
+}
+#endif
+
+static int airoc_mgmt_ap_disable(const struct device *dev)
+{
+	int ret = 0;
+	struct airoc_wifi_data *data = (struct airoc_wifi_data *)dev->data;
+
+	if (k_sem_take(&data->sema_common, K_MSEC(AIROC_WIFI_WAIT_SEMA_MS)) != 0) {
+		return -EAGAIN;
+	}
+
+	whd_wifi_deregister_event_handler(airoc_ap_if, ap_event_handler_index);
+	whd_wifi_stop_ap(airoc_ap_if);
```

Comment:
Both whd_wifi_deregister_event_handler and whd_wifi_stop_ap have return codes, so we should at least assert they are successful. Given that we are disabling the AP, there isn't much that can be done if they fail, so maybe we don't need to return the values from these functions in a fail state.